aten/src/ATen/native/mps/operations/Indexing.mm

```diff
@@ -375,15 +393,16 @@ Tensor nonzero_fallback(const Tensor& self) {
     };
 
     NSDictionary<MPSGraphTensor*, MPSGraphTensorData*>* results = @{
-      outputPlaceholder.getMPSGraphTensor() : outputPlaceholder.getMPSGraphTensorData()
+      outputPlaceholder.getMPSGraphTensor() : outputPlaceholder.getMPSGraphTensorData(),
+      countNonzeroPlaceholder.getMPSGraphTensor() : countNonzeroPlaceholder.getMPSGraphTensorData()
     };
 
     runMPSGraph(stream, cachedGraph->graph(), feeds, results);
-    if (!contiguous_output) {
-      out_.copy_(out);
-    }
   }
 
+  int32_t total_nonzero = count_nonzero.item<int32_t>();
```

Comment:
This will trigger two hard-syncs, one from item, and one for the following `out_.copy_()`.
Later (not now), I suggest we create a dedicated MPSGraph for this part. We pre-allocate `out_` with the same size of self (so we don't overflow the buffer when resizing), and do the zero-counting and resizing of output in a single MPSGraph op. We can discuss that later.

Reply:
> We pre-allocate out_ with the same size of self (so we don't overflow the buffer when resizing)

Doing that might allocate a new buffer and change the pointer of the out buffer (causing a failure in the test).
E.g in case the user has a pre-allocated buffer from a previous nonzero op (and they know the exact number of nonzeros) doing again `nonzero(input, out=preallocated_out)` and resizing the output at the beginning of the function to match the `input`, could allocate new memory if input's number of elements is larger than output's number of elements ([resize_](https://pytorch.org/docs/stable/generated/torch.Tensor.resize_.html)]. 
Previously the op it was calling into `count_nonzero` at the beggining of the function to get the number of elements, this change makes it to get the number of elements from the same graph as `nonzero` (seemed a little bit faster when testing compared to previous method).

> and do the zero-counting and resizing of output in a single MPSGraph op. We can discuss that later.

We can do it in the graph (both nonzero and count_nonzero happen in the same graph now), but the returned Tensor's shape we've preallocated in the beginning (not the MPSGraphTensor*) would still be wrong and we'd need to sync at the end (the `.item()` part) to get the number of nonzeros and resize it correctly. And if we've preallocated the output at the beginning it would hit the issue from above (it would work for 99% of the tests to preallocate output in the beginning, but would fail were they're passing a preallocated output to us)